### PR TITLE
Reverting back to django-celery background result, refactored code, created celery_app, reorganized code

### PR DESCRIPTION
**Pull request by alalamin25**
You might need to run python manage.py collectstatics for ckeditor

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,4 +4,6 @@ __pycache__/
 wsgi.py
 db.sqlite3
 media/
-static/common/google_analytics.js
\ No newline at end of file
+static/common/google_analytics.jsbash_commands/
```
Comment (tareqobaida) [inline .gitignore]: why did you remove google_analytics.js from .gitignore?

Comment (alalamin25) [inline .gitignore]: No. I even rechecked now. It's in the .gitignore file

Comment (tareqobaida) [inline .gitignore]: strange, in my repo also it's showing correctly but in change view it's showing changed.

Comment (alalamin25) [inline .gitignore]: There is a ending sland at the end of static/common/google_analytics.jsbash_commands/
should it be static/common/google_analytics.jsbash_commands

File: settings/local.py [hunk 2]
```diff
@@ -27,3 +27,8 @@
         },
     },
 }
+
+
```
Comment (tareqobaida) [inline settings/local.py]: You don't need to commit this file if you are using it just for your own test

File: skype_consultancy/tasks.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    # This is just a helper function to send email async.
+    # See the output in the worker process console
+    print("\n\n Goint to send email asynconously\n\n")
+    return send_mail(subject, body_email, to_email)
+
+
+@shared_task
+def send_skype_email_before_hour(event_id, hour=12):
+    # Change and updat the body_email make it more customized
+    print("\n\nGoing to send email %d hour before " % hour)
```
Comment (tareqobaida) [inline skype_consultancy/tasks.py]: use logging

File: skype_consultancy/tasks.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+@shared_task
+def send_mail_async(subject, body_email, to_email):
+    # This is just a helper function to send email async.
+    # See the output in the worker process console
+    print("\n\n Goint to send email asynconously\n\n")
+    return send_mail(subject, body_email, to_email)
+
+
+@shared_task
```
Comment (tareqobaida) [inline skype_consultancy/tasks.py]: Looks like you can combine all those functions in a single function

Comment (tareqobaida) [inline skype_consultancy/tasks.py]: Remember, don't repeat your code

File: skype_consultancy/tasks.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                     Thanks,
+                     Support Team
+                     My Study Notebook
+                     """.format(reg_user.attendee.first_name, event.title,
+                                event.start_time, event.duration)
+        send_mail(subject, body_email, to_email)
+
+
+@shared_task
+def skype_event_group_email(skype_email_id):
```
Comment (tareqobaida) [inline skype_consultancy/tasks.py]: you can move this piece of code to the save_model method.

File: skype_consultancy/tasks.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    print(registered_user_list)
+    for reg_user in registered_user_list:
+        print("\n\n group email for loop\n")
+        to_email = [reg_user.attendee.email, ]
+        subject = skype_email.email_subject
+        body_email = skype_email.email_body
+        send_mail(subject, body_email, to_email)
+
+# This task is for practice purpose
+@shared_task
```
Comment (tareqobaida) [inline skype_consultancy/tasks.py]: remove unnecessary codes before committing

File: MSNB/celery.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+)
+app.conf.update(
+    CELERY_RESULT_BACKEND='djcelery.backends.cache:CacheBackend',
+)
+
+app.conf.update(
+    CELERY_TASK_SERIALIZER='json',
+    CELERY_ACCEPT_CONTENT=['json'],  # Ignore other content
+    CELERY_RESULT_SERIALIZER='json',
+    CELERY_TIMEZONE='Europe/Oslo',
```
Comment (alalamin25) [inline MSNB/celery.py]: texas has two time zone one is UTC -6/-5
which one to use

File: event/admin.py [hunk 8]
```diff
@@ -1,15 +1,85 @@
...
 # Register your models here.
+
+
 class RegistrationInline(admin.TabularInline):
     model = Registration
 
 
 class EventAdmin(admin.ModelAdmin):
     inlines = [RegistrationInline]
+    raw_id_fields = ('presenter',)
```
Comment (alalamin25) [inline event/admin.py]: presenter is a foreign key from user model to event model. So when you create new event you will have to choose a presenter(User model object). 
By default you will have to choose the presenter user from that drop down menu. The problem is if you have 50000 users then it will be very very difficult to choose one user from that drop down menu. But luckily with this little tweak you will be able to search and find the current user from this 50000 users very easily and add him as a presenter to Event model

File: faq/tests/test_views.py [hunk 9]
```diff
@@ -1,48 +1,55 @@
...
+        if (Question.objects.all().count() > 0):
+            self.assertIsInstance(
+                response.context['recent_q'], models.Question)
+        # question = Question.objects.create(text="new question")
+
+
+class TestFaqSearchResult(TestCase):
 
-class TestRelatedModels(TestCase):
     def setUp(self):
```
Comment (alalamin25) [inline faq/tests/test_views.py]: I did not need anything to setUp so I skipped this.
should i write:
def setUp(self):
    pass

File: home/tests/test_views.py [hunk 10]
```diff
@@ -1,16 +1,65 @@
...
+
+
+class TestHomeView(TestCase):
+    
+    def test_status_code(self):
+        response = self.client.get(reverse("home"))
+        self.assertEqual(response.status_code, 200)
+
+
+class TestDecesionMakingView(TestCase):
```
Comment (alalamin25) [inline home/tests/test_views.py]: I tried to follow the TDD book.
I actually download the source code of the book and tried to follow.
In that book for every method in views.py there were a Class so I added them like that.
Because I might write additional tests for those classes too. Again if the views.py 's methods get complicated then additional tests can very easily be added to as there is a seperate class for every views method

File: skype_consultancy/tasks.py (hunk 5 above)
Comment (alalamin25) [inline skype_consultancy/tasks.py]: If you have 10 users registered in that event then if you try to send email to them from admin panel. Then when you click save it will take a long time to save. Because until all the events are saved the page will not load. You can try to send email asynchronously  by calling send_mail_async method from admin.py but In that case you are  still calling a method from tasks.py

File: skype_consultancy/tasks.py (hunk 4 above)
Comment (alalamin25) [inline skype_consultancy/tasks.py]: I have created two method in tasks.py send_remainder_email method and send_feedback_remainder_email
They can easily be merged into one.
Then you can just call them by Editing subject and body parameter(Remainder_email and feedback_remainder emails’ subject and body will be different.)
I created two methods for remainder and feedback remainder because if later in future you want might want to do something extra while sending remainder_email. In that case you can simply add those functionalities to the to the remainder_email method as well as sending_email for feedback.
If you want me to merge then I can merge them

File: faq/tests/test_views.py [hunk 11]
```diff
@@ -1,48 +1,55 @@
...
+    def test_faq_view_renders_faq_template(self):
+        response = self.client.get('/faq/')
+        self.assertTemplateUsed(response, 'faq/faq.html')
+
+    def test_faq_page_uses_faq_search_form(self):
+        response = self.client.get('/faq/')
+        self.assertIsInstance(
+            response.context['faq_search_form'], FaqSearchForm)
+
+    def test_faq_page_uses_question_model(self):
```
Comment (alalamin25) [inline faq/tests/test_views.py]: Template system is complex. It has multiple inheritance. This test is just to make sure template is loading up the form.

File: celery_app/models.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+from django.db import models
+
+
+# All background task_id will saved here.
+# So in case any scheduled task needs to be revoked then we'll find the task_id
+# in this model
+
+class TaskList(models.Model):
+    task_name = models.CharField(max_length=50)
+    task_id = models.CharField(max_length=50)
```
Comment (alalamin25) [inline celery_app/models.py]: Actually two fields will be necessary.
task_name is necessary to find scheduled tasks of of skype_event_1 or skype_event_2
task_id is necessary to find all the scheduled task under that task. If a task had only one task then task_id were not necessary.
EX:
for skype_event_1(id=1)
task_name is event1
and this event will have 3 scheduled task those different tasks with 3 task_id
this task_id field will store those id.

Actually TaskList:
taskname: foreignkey(event)
task_id : foreignkey(celery_task_id)

task_name is not added as foreignkey(event) but added as string because then TaskList will be specific to only store event tasks. if we want to store this taskList to store other events/tasks then we can easily do that.

File: skype_consultancy/tasks.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                     Support Team
+                     My Study Notebook
+                     """.format(reg_user.attendee.first_name, event.title,
+                                event.start_time, event.duration)
+        send_mail(subject, body_email, to_email)
+
+
+# Remainder Email and Feedback remainder's emails subject body will be
+# significantly different. Thats why there are two methods to edit and
+# customize these two type email
```
Comment (tareqobaida) [inline skype_consultancy/tasks.py]: or even you don't need to define those functions, just simply use send_mail_async

Comment (tareqobaida) [inline skype_consultancy/tasks.py]: You can use a separate file for constants like email body

File: skype_consultancy/tasks.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
+from __future__ import absolute_import
```
Comment (tareqobaida) [inline skype_consultancy/tasks.py]: This file should be inside celery_app folder

File: event/admin.py [hunk 15]
```diff
@@ -1,15 +1,108 @@
...
+
+    # Here the save model is going to be overwritten. When someone creates or edits
+    # edits a event this method is going to be called. When new event is created or
+    # existing event is edited the background email scheduling is going to be
+    # created or updated too. Here that background task is going to be created
+    # or updated.
+
+    # going to make a schedule to send email before 12 hours by synchoneously
```
Comment (alalamin25) [inline event/admin.py]: yes!

File: requirements.txt [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+celery==3.1.24
+coverage==4.2
+defusedxml==0.4.1
+Django==1.8
+django-celery==3.1.17
+django-ckeditor==5.1.1
+django-compat==1.0.13
+django-debug-toolbar==1.5
+flower==0.9.1
+jsonfield==1.0.3
```
Comment (alalamin25) [inline requirements.txt]: flower is for debugging purpose of celery
I did not installed other package. They were probably in pip

File: event/admin.py [hunk 17]
```diff
@@ -1,15 +1,108 @@
...
+    # calling send_remainder_email
+    def schedule_first_remainder_email(self, obj, task_name):
```
Comment (alalamin25) [inline event/admin.py]: Actually I created these methods to take care of calculation of time_delta and expiry_time_delta
This calculations looks ugly. So i put them in another method to calculate. If i create only one method these time calculation will have to done before.
And doesn't it look simpler that all the logic and calculation of remainder_email is in one place

File: skype_consultancy/tasks.py (hunk 14 above)
Comment (alalamin25) [inline skype_consultancy/tasks.py]: okay. But these tasks are related to skype_consultancy. Shouldn't each app's tasks reside in it's app

File: MSNB/celery.py (hunk 7 above)
Comment (alalamin25) [inline MSNB/celery.py]: Your django timezone is UTC and celery timezone is UTC-6.
So to use the scheduled task properly you'll have to calculate properly.
Do you really want to have two different timezone?

File: event/admin.py (hunk 17 above)
Comment (alalamin25) [inline event/admin.py]: Admin.py is not supposed to handing all these logic. Its main purpose is to customize admin site.
So we can put the send_first_remainder, second_remainder, feed_back_email methods to another file and make our admin.py cleaner

File: home/tests/test_views.py [hunk 18]
```diff
@@ -1,16 +1,65 @@
...
+
+
+class TestApplicationView(TestCase):
+    
+    def test_application_status_code(self):
+        response = self.client.get(reverse("application"))
+        self.assertEqual(response.status_code, 200)
+
+
+class TestGoogleSearchView(TestCase):
```
Comment (alalamin25) [inline home/tests/test_views.py]: As I tried to write test for search page. The google custom search does not work. It either does not provide any result or even if it does the result is not remotely relevant.
Would you please check that feature?

File: celery_app/models.py (hunk 12 above)
Comment (tareqobaida) [inline celery_app/models.py]: I got your point. In that case please change the field name 'task_name' to something like 'parent_task' and 'task_id' to something like 'celery_task_id'.

File: requirements.txt (hunk 16 above)
Comment (alalamin25) [inline requirements.txt]: I did not install all those packages. As you suggested perhaps one package depended on other and so was installed automatically. So I do not know know which one to keep and which one to delete?

**Comment by tareqobaida** [comment]
Tested celery in local environment. It's working as expected. Need to change time zone to TIME_ZONE = 'US/Central'
